File: frontend/src/types/job.ts
/**
 * Job-related types for the jobs workflow.
 */
import type { LinkedEmailContext } from "./email";

/**
 * All possible job statuses.
 * Order matters - this is the linear flow order.
 */
export const JOB_STATUSES = [
  "new",
  "analyzed",
  "prepped",
  "reviewed",
  "applied",
  "interviewing",
  "rejected",
] as const;

/**
 * Status of a job in the user's workflow.
 */
export type JobStatus = (typeof JOB_STATUSES)[number];

export const PRE_APPLIED_JOB_STATUSES: JobStatus[] = ["new", "analyzed", "prepped", "reviewed"];

export const POST_APPLIED_JOB_STATUSES: JobStatus[] = ["applied", "interviewing", "rejected"];

export const DEFAULT_JOB_STATUS_FILTERS: JobStatus[] = [...PRE_APPLIED_JOB_STATUSES];

/**
 * Metadata for each job status including display info and allowed transitions.
 */
export const JOB_STATUS_CONFIG: Record<
  JobStatus,
  {
    label: string;
    description: string;
    /** Statuses that can transition TO this status */
    allowedFrom: JobStatus[];
  }
> = {
  new: {
    label: "New",
    description: "Added and waiting for application requirements",
    allowedFrom: [], // Initial state, nothing transitions to new
  },
  analyzed: {
    label: "Analyzed",
    description: "Application requirements captured and ready for prep",
    allowedFrom: [],
  },
  prepped: {
    label: "Prepped",
    description: "Prep notes and optional application materials generated",
    allowedFrom: ["analyzed"],
  },
  reviewed: {
    label: "Reviewed",
    description: "Ready to apply",
    allowedFrom: ["prepped"],
  },
  applied: {
    label: "Applied",
    description: "Application submitted",
    allowedFrom: ["new", "analyzed", "prepped", "reviewed"],
  },
  interviewing: {
    label: "Interviewing",
    description: "In interview process",
    allowedFrom: ["applied"],
  },
  rejected: {
    label: "Rejected",
    description: "Application was declined",
    allowedFrom: ["applied", "interviewing"],
  },
};

/**
 * Check if a status transition is allowed.
 */
export function canTransitionTo(from: JobStatus, to: JobStatus): boolean {
  if (from === to) return false;
  return JOB_STATUS_CONFIG[to].allowedFrom.includes(from);
}

/**
 * Get all statuses that the given status can transition to.
 */
export function getAllowedTransitions(from: JobStatus): JobStatus[] {
  return JOB_STATUSES.filter((status) => canTransitionTo(from, status));
}

/**
 * How a job was discovered/added to the system.
 */
export type IngestionSource = "scrape" | "email" | "manual" | "openclaw";

/**
 * A job listing from the database.
 */
export interface Job {
  id: string;
  user_id: string;
  profile_id: string | null;
  title: string;
  company: string;
  location: string | null;
  description: string | null;
  job_url: string;
  salary_range: string | null;
  date_posted: string | null;
  source: string | null;
  ingestion_source: IngestionSource | null;
  relevance_score: number | null;
  reasoning: string | null;
  status: JobStatus;
  search_terms: string | null;
  notes: string | null;
  // Additional scrape fields
  is_remote: boolean | null;
  job_type: string | null;
  company_url: string | null;
  // Prep materials
  cover_letter: string | null;
  cover_letter_file_path: string | null;
  cover_letter_generated_at: string | null;
  prep_notes: string | null;
  prepped_at: string | null;
  application_type: "easy_apply" | "ats" | "direct" | "email" | "unknown" | null;
  application_url: string | null;
  requires_cover_letter: boolean | null;
  cover_letter_requested?: boolean | null;
  requires_resume: boolean | null;
  detected_fields: Record<string, unknown> | null;
  screening_questions: Array<Record<string, unknown>> | null;
  screening_answers: Record<string, string> | null;
  ats_family?: string | null;
  analysis_source?: string | null;
  analyzed_at: string | null;
  has_application_analysis?: boolean;
  is_prep_eligible?: boolean;
  applied_at: string | null;
  application_method: string | null;
  confirmation_code: string | null;
  linked_email: LinkedEmailContext | null;
  created_at: string;
  updated_at: string | null;
}

/**
 * True when the job currently has real cover-letter text that can be rendered to PDF.
 */
export function hasCoverLetterText(coverLetter: string | null | undefined): boolean {
  const text = coverLetter?.trim();
  return Boolean(text);
}

/**
 * Reviewed jobs only need PDF generation when a real cover letter exists or
 * application analysis explicitly says one is required.
 */
export function shouldGenerateReviewPdf(
  job: Pick<Job, "cover_letter" | "requires_cover_letter"> & {
    cover_letter_requested?: boolean | null;
  },
  draftCoverLetter?: string | null
): boolean {
  if (hasCoverLetterText(draftCoverLetter ?? job.cover_letter)) {
    return true;
  }

  return job.requires_cover_letter === true || job.cover_letter_requested === true;
}

export function getScreeningQuestionText(question: Record<string, unknown>): string {
  const candidates = [
    question.question,
    question.label,
    question.prompt,
    question.name,
    question.text,
  ];

  const match = candidates.find(
    (value): value is string => typeof value === "string" && value.trim().length > 0
  );

  return match?.trim() ?? "";
}

export function getOrderedScreeningAnswers(
  job: Pick<Job, "screening_questions" | "screening_answers">
): Array<{ question: string; answer: string }> {
  const orderedQuestions =
    job.screening_questions?.map(getScreeningQuestionText).filter(Boolean) ?? [];
  const answers = job.screening_answers ?? {};
  const seen = new Set<string>();
  const ordered = orderedQuestions.flatMap((question) => {
    const answer = answers[question];
    if (!answer) {
      return [];
    }
    seen.add(question);
    return [{ question, answer }];
  });

  const remaining = Object.entries(answers)
    .filter(([question, answer]) => Boolean(question.trim()) && Boolean(answer.trim()))
    .filter(([question]) => !seen.has(question))
    .map(([question, answer]) => ({ question, answer }));

  return [...ordered, ...remaining];
}

/**
 * Abbreviated job info for lists.
 */
export interface JobSummary {
  id: string;
  title: string;
  company: string;
  location: string | null;
  relevance_score: number | null;
  status: JobStatus;
  job_url: string;
}

/**
 * Response from listing jobs.
 */
export interface JobListResponse {
  jobs: Job[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

/**
 * Statistics about user's jobs.
 */
export interface JobStats {
  total: number;
  new: number;
  analyzed: number;
  prepped: number;
  reviewed: number;
  applied: number;
  interviewing: number;
  rejected: number;
  avg_score: number | null;
  high_scoring: number;
}

/**
 * Filters for querying jobs.
 */
export interface JobFilters {
  status?: JobStatus;
  statuses?: JobStatus[];
  source?: string;
  ingestion_source?: IngestionSource;
  min_score?: number;
  max_score?: number;
  search?: string;
  prep_eligible?: boolean;
  posted_within_hours?: number;
  page?: number;
  page_size?: number;
  sort_by?: "created_at" | "relevance_score" | "date_posted" | "company";
  sort_order?: "asc" | "desc";
}

/**
 * Request to update a job.
 */
export interface JobUpdate {
  status?: JobStatus;
  notes?: string;
  cover_letter?: string;
  prep_notes?: string;
}

export interface ManualAnalyzeRequest {
  requires_cover_letter?: boolean;
  screening_questions?: string[];
}

export interface ManualJobCreateRequest {
  title: string;
  company: string;
  job_url: string;
  location?: string | null;
  description?: string | null;
  salary_range?: string | null;
  date_posted?: string | null;
  source?: string | null;
  is_remote?: boolean | null;
  job_type?: string | null;
  company_url?: string | null;
  profile_id?: string | null;
  notes?: string | null;
}

// ===========================================================================
// Resume Types
// ===========================================================================

/**
 * A user's uploaded resume file.
 */
export interface Resume {
  id: string;
  user_id: string;
  name: string;
  original_filename: string;
  file_size: number;
  mime_type: string;
  is_primary: boolean;
  has_text: boolean;
  created_at: string;
  updated_at: string | null;
}

/**
 * Summary of resume for lists and selectors.
 */
export interface ResumeSummary {
  id: string;
  name: string;
  original_filename: string;
  is_primary: boolean;
  has_text: boolean;
}

/**
 * Resume info embedded in profile response.
 */
export interface ResumeInfo {
  id: string;
  name: string;
  original_filename: string;
  has_text: boolean;
}

// ===========================================================================
// Job Profile Types
// ===========================================================================

/**
 * Story info embedded in profile response.
 */
export interface StoryInfo {
  id: string;
  name: string;
}

/**
 * Project info embedded in profile response.
 */
export interface ProjectInfo {
  id: string;
  name: string;
  has_text: boolean;
}

/**
 * Job profile with resume, story, and projects link and preferences.
 * A user can have multiple profiles with different configurations.
 */
export interface JobProfile {
  id: string;
  user_id: string;
  name: string;
  is_default: boolean;
  resume_id: string | null;
  resume: ResumeInfo | null;
  story_id: string | null;
  story: StoryInfo | null;
  project_ids: string[] | null;
  projects: ProjectInfo[] | null;
  target_roles: string[] | null;
  target_locations: string[] | null;
  min_score_threshold: number;
  preferences: Record<string, unknown> | null;
  // Contact info for cover letters
  contact_full_name: string | null;
  contact_phone: string | null;
  contact_email: string | null;
  contact_location: string | null;
  contact_website: string | null;
  created_at: string;
  updated_at: string | null;
}

/**
 * Summary of profile for lists.
 */
export interface JobProfileSummary {
  id: string;
  name: string;
  is_default: boolean;
  has_resume: boolean;
  has_cover_letter_full_name: boolean;
  resume_name: string | null;
  has_story: boolean;
  story_name: string | null;
  project_count: number;
  target_roles_count: number;
  min_score_threshold: number;
}

/**
 * Request to create a profile.
 */
export interface JobProfileCreate {
  name: string;
  is_default?: boolean;
  resume_id?: string | null;
  story_id?: string | null;
  project_ids?: string[] | null;
  target_roles?: string[] | null;
  target_locations?: string[] | null;
  min_score_threshold?: number;
  preferences?: Record<string, unknown> | null;
  // Contact info for cover letters
  contact_full_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  contact_location?: string | null;
  contact_website?: string | null;
}

/**
 * Partial update to profile.
 */
export interface JobProfileUpdate {
  name?: string;
  is_default?: boolean;
  resume_id?: string | null;
  story_id?: string | null;
  project_ids?: string[] | null;
  target_roles?: string[] | null;
  target_locations?: string[] | null;
  min_score_threshold?: number;
  preferences?: Record<string, unknown> | null;
  // Contact info for cover letters
  contact_full_name?: string | null;
  contact_phone?: string | null;
  contact_email?: string | null;
  contact_location?: string | null;
  contact_website?: string | null;
}

// ===========================================================================
// Story Types
// ===========================================================================

/**
 * A user's story/narrative text.
 */
export interface Story {
  id: string;
  user_id: string;
  name: string;
  content: string;
  is_primary: boolean;
  created_at: string;
  updated_at: string | null;
}

/**
 * Summary of story for lists and selectors.
 */
export interface StorySummary {
  id: string;
  name: string;
  is_primary: boolean;
  content_preview: string;
}

/**
 * Request to create a story.
 */
export interface StoryCreate {
  name: string;
  content: string;
  is_primary?: boolean;
}

/**
 * Partial update to story.
 */
export interface StoryUpdate {
  name?: string;
  content?: string;
  is_primary?: boolean;
}

// ===========================================================================
// Project Types
// ===========================================================================

/**
 * A user's project description.
 */
export interface Project {
  id: string;
  user_id: string;
  name: string;
  original_filename: string;
  file_size: number;
  mime_type: string;
  has_text: boolean;
  created_at: string;
  updated_at: string | null;
}

/**
 * Summary of project for lists.
 */
export interface ProjectSummary {
  id: string;
  name: string;
  original_filename: string;
  has_text: boolean;
}

/**
 * Project text content response.
 */
export interface ProjectTextResponse {
  id: string;
  name: string;
  text_content: string | null;
}

/**
 * Partial update to project.
 */
export interface ProjectUpdate {
  name?: string;
}

// ===========================================================================
// Pipeline Error Types
// ===========================================================================

/**
 * Structured error returned when a job profile is required but not available.
 * The frontend uses this to show a profile selector instead of a generic error.
 */
export interface ProfileRequiredError {
  error_type: "profile_required";
  message: string;
  available_profiles: JobProfileSummary[];
  create_profile_url: string;
}

File: backend/app/schemas/job.py
"""Job schemas for API request/response handling."""

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import Field, model_validator

from app.db.models.job import JobStatus
from app.schemas.base import BaseSchema, TimestampSchema
from app.schemas.linked_email import LinkedEmailContext

# Application type literals
ApplicationType = Literal["easy_apply", "ats", "direct", "email", "unknown"]
IngestionSource = Literal["scrape", "email", "manual", "openclaw"]


class JobBase(BaseSchema):
    """Base job schema with common fields."""

    title: str = Field(max_length=500, description="Job title")
    company: str = Field(max_length=255, description="Company name")
    location: str | None = Field(default=None, max_length=255, description="Job location")
    description: str | None = Field(default=None, description="Job description")
    job_url: str = Field(max_length=2048, description="URL to job posting")
    salary_range: str | None = Field(
        default=None, max_length=100, description="Salary range if available"
    )
    date_posted: datetime | None = Field(default=None, description="When the job was posted")
    source: str | None = Field(
        default=None, max_length=50, description="Source (linkedin, indeed, etc.)"
    )
    ingestion_source: IngestionSource | None = Field(
        default=None,
        max_length=20,
        description="How job was discovered (scrape, email, manual, openclaw)",
    )
    is_remote: bool | None = Field(default=None, description="Whether the job is remote")
    job_type: str | None = Field(
        default=None,
        max_length=50,
        description="Job type (fulltime, parttime, internship, contract)",
    )
    company_url: str | None = Field(
        default=None, max_length=2048, description="URL to company page"
    )


class JobCreate(JobBase):
    """Schema for creating a job (internal use by pipeline)."""

    profile_id: UUID | None = Field(
        default=None,
        description="Optional profile to reuse later as prep context",
    )
    relevance_score: float | None = Field(
        default=None, ge=0.0, le=10.0, description="AI-computed relevance score (0-10)"
    )
    reasoning: str | None = Field(default=None, description="AI reasoning for the score")
    search_terms: str | None = Field(
        default=None, max_length=500, description="Search terms used to find this job"
    )


class ManualJobCreateRequest(JobBase):
    """Schema for manually creating a job with optional prep-profile context."""

    profile_id: UUID | None = Field(
        default=None,
        description="Optional profile to associate for later prep",
    )
    notes: str | None = Field(default=None, description="Optional notes for the job")


class ManualAnalyzeRequest(BaseSchema):
    """Schema for manually marking a job ready for prep."""

    requires_cover_letter: bool = Field(
        default=False,
        description="Whether this application needs a cover letter",
    )
    screening_questions: list[str] = Field(
        default_factory=list,
        description="Optional custom questions to prepare answers for",
    )


class JobUpdate(BaseSchema):
    """Schema for updating a job.

    Users can update status, notes, and prep materials.
    """

    status: JobStatus | None = Field(default=None, description="Job status in workflow")
    notes: str | None = Field(default=None, description="User notes about the job")
    cover_letter: str | None = Field(default=None, description="Generated cover letter")
    prep_notes: str | None = Field(
        default=None, description="Prep notes (highlights + talking points)"
    )


class JobResponse(JobBase, TimestampSchema):
    """Schema for reading a job (API response)."""

    id: UUID
    user_id: UUID
    profile_id: UUID | None = None
    relevance_score: float | None = None
    reasoning: str | None = None
    status: JobStatus = JobStatus.NEW
    search_terms: str | None = None
    notes: str | None = None
    # Prep materials
    cover_letter: str | None = None
    cover_letter_file_path: str | None = None
    cover_letter_generated_at: datetime | None = None
    prep_notes: str | None = None
    prepped_at: datetime | None = None
    # Application analysis
    application_type: ApplicationType | None = None
    application_url: str | None = None
    requires_cover_letter: bool | None = None
    cover_letter_requested: bool | None = None
    requires_resume: bool | None = None
    detected_fields: dict[str, Any] | None = None
    screening_questions: list[dict[str, Any]] | None = None
    screening_answers: dict[str, str] | None = None
    ats_family: str | None = None
    analysis_source: str | None = None
    analyzed_at: datetime | None = None
    has_application_analysis: bool = False
    is_prep_eligible: bool = False
    # Application submission
    applied_at: datetime | None = None
    application_method: str | None = None
    confirmation_code: str | None = None
    # Linked email context
    linked_email: LinkedEmailContext | None = None

    @model_validator(mode="before")
    @classmethod
    def _resolve_linked_email(cls, data: Any) -> Any:
        """Build linked_email from the source_email_message relationship."""
        if isinstance(data, dict):
            return data
        msg = getattr(data, "source_email_message", None)
        if msg is not None and not hasattr(data, "linked_email"):
            source = getattr(msg, "source", None)
            object.__setattr__(
                data,
                "linked_email",
                LinkedEmailContext(
                    id=msg.id,
                    source_email_address=source.email_address if source else "",
                    gmail_message_id=msg.gmail_message_id,
                    gmail_thread_id=msg.gmail_thread_id,
                    subject=msg.subject,
                    from_address=msg.from_address,
                    received_at=msg.received_at,
                    bucket=msg.bucket,
                    summary=msg.summary,
                ),
            )
        return data


class JobSummary(BaseSchema):
    """Abbreviated job info for lists and pipeline output."""

    id: UUID
    title: str
    company: str
    location: str | None = None
    relevance_score: float | None = None
    status: JobStatus = JobStatus.NEW
    job_url: str
    source: str | None = None
    ingestion_source: str | None = None


class JobListResponse(BaseSchema):
    """Paginated list of jobs."""

    jobs: list[JobResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class JobStatsResponse(BaseSchema):
    """Statistics about user's jobs."""

    total: int = 0
    new: int = 0
    analyzed: int = 0
    prepped: int = 0
    reviewed: int = 0
    applied: int = 0
    interviewing: int = 0
    rejected: int = 0
    avg_score: float | None = None
    high_scoring: int = 0  # Jobs with score >= 7.0


class JobFilters(BaseSchema):
    """Filters for querying jobs."""

    status: JobStatus | None = None
    statuses: list[JobStatus] | None = None
    source: str | None = None
    ingestion_source: IngestionSource | None = Field(
        default=None,
        description="Filter by how job was discovered (scrape, email, manual, openclaw)",
    )
    min_score: float | None = Field(default=None, ge=0.0, le=10.0)
    max_score: float | None = Field(default=None, ge=0.0, le=10.0)
    search: str | None = Field(default=None, description="Search in title, company, description")
    prep_eligible: bool | None = Field(
        default=None,
        description="Filter to jobs explicitly analyzed and eligible for prep",
    )
    posted_within_hours: int | None = Field(
        default=None,
        ge=1,
        description="Filter to jobs posted within the last N hours (e.g., 24, 48, 72)",
    )
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    sort_by: Literal["created_at", "relevance_score", "date_posted", "company"] = "created_at"
    sort_order: Literal["asc", "desc"] = "desc"
